cdap-kubernetes/src/main/java/io/cdap/cdap/k8s/common/AbstractWatcherThread.java

```diff
@@ -184,42 +201,178 @@ public final void run() {
+    }
+  }
+
+  /**
+   * Creates a new {@link Watchable} for watching for changes.
+   *
+   * @param lastResourceVersion if provided, the watch will start from this version. If it is {@code null},
+   *                            a list of the latest resources will be fetch first to compute the difference with
+   *                            the previously known set and have the watch starting from there.
+   * @return a {@link Watchable}
+   */
+  private Watchable<DynamicKubernetesObject> createWatchable(@Nullable String lastResourceVersion)
+    throws IOException, ApiException {
+
+    ApiClient apiClient = getApiClient();
+    DynamicKubernetesApi api = new DynamicKubernetesApi(group, version, plural, apiClient);
+
+    ListOptions options = new ListOptions();
+    updateListOptions(options);
+
+    String resourceVersion = lastResourceVersion;
+
+    if (resourceVersion == null) {
+      KubernetesApiResponse<DynamicKubernetesListObject> listResult = api.list(namespace, options);
+
+      // Throw exception if the list call failed
+      listResult.throwsApiException();
+
+      DynamicKubernetesListObject listObject = listResult.getObject();
+      resourceVersion = listObject.getMetadata().getResourceVersion();
+
+      LOG.debug("Fetched '{}/{}/{}' list with resource version as {}", group, version, plural, resourceVersion);
+
+      Map<String, T> cachedResources = changeListener.getCachedResources();
+      Set<String> currentSet = new HashSet<>(cachedResources.keySet());
+
+      // Add existing resources
+      for (DynamicKubernetesObject obj : listObject.getItems()) {
+        String name = obj.getMetadata().getName();
+        T resource = decodeResource(obj);
+
+        // If the resource is not known before, emit an add event
+        if (!currentSet.remove(name)) {
+          changeListener.resourceAdded(resource);
+        } else if (!Objects.equals(resource, cachedResources.get(name))) {
+          // Otherwise, if there is changes in the resource, emit a modified event
+          changeListener.resourceModified(resource);
+        }
+      }
+
+      // Emit deleted event for resources that were deleted
+      for (String deleted : currentSet) {
+        T resource = cachedResources.get(deleted);
+        if (resource != null) {
+          changeListener.resourceDeleted(resource);
+        }
+      }
+    }
+
+    LOG.debug("Start watching '{}/{}/{}' starting at resource version {}", group, version, plural, resourceVersion);
+
+    // Create the new watch
+    options = new ListOptions();
+    updateListOptions(options);
+    options.setResourceVersion(resourceVersion);
+    Watchable<DynamicKubernetesObject> watch = wrapWatchableClose(api.watch(namespace, options));
+    Closeables.close(this.watch.getAndSet(watch), true);
+    return watch;
+  }
+
+  private T decodeResource(DynamicKubernetesObject object) throws IOException {
+    return getApiClient().getJSON().getGson().fromJson(object.getRaw(), resourceType);
```

Comment:
Just for my understanding, why do we need to use the api client to decode the resource? Wouldn't something like GSON.fromJson(object.getRaw(), resourceType) do the same thing?

Reply:
The `Gson` object inside the ApiClient is configured with `TypeAdapter` for decoding k8s object to correct types.